packages/smithy/smithy/lib/src/http/http_operation.dart

```diff
@@ -0,0 +1,378 @@
+  }
+
+  /// Builds the HTTP request for the given [input].
+  HttpRequest buildRequest(Input input);
+
+  /// Builds the output from the [payload] and metadata from the HTTP
+  /// [response].
+  Output buildOutput(
+    // This is (kind of) a hack to allow `OutputPayload` to always be non-null
+    // even if the payload type is nullable. We need the non-null version to
+    // interop with built_value correctly.
+    covariant Object? payload,
+    AWSStreamedHttpResponse response,
+  );
+
+  /// The protocols used by this operation for all serialization/deserialization
+  /// of wire formats.
+  Iterable<HttpProtocol<InputPayload, Input, OutputPayload, Output>>
+      get protocols;
+
+  @override
+  List<SmithyError> get errorTypes;
+
+  /// The success code for the operation.
+  ///
+  /// Accepts the operation output since some output types embed the success
+  /// code to allow for dynamic success codes.
+  int successCode([Output? output]);
+
+  /// The number of times the operation has been retried.
+  @visibleForTesting
+  int debugNumRetries = 0;
+
+  /// The base URI for the operation.
+  Uri get baseUri;
+
+  /// The endpoint for the operation.
+  Endpoint get endpoint => Endpoint(uri: baseUri);
+
+  /// The retry handler for the operation.
+  Retryer get retryer => const Retryer();
+
+  @visibleForTesting
+  HttpProtocol<InputPayload, Input, OutputPayload, Output> resolveProtocol({
+    ShapeId? useProtocol,
+  }) {
+    return useProtocol == null
+        ? protocols.first
+        : protocols.firstWhere(
+            (el) => el.protocolId == useProtocol,
+            orElse: () => protocols.first,
+          );
+  }
+
+  @visibleForTesting
+  Future<AWSStreamedHttpRequest> createRequest(
+    HttpRequest request,
+    HttpProtocol<InputPayload, Input, OutputPayload, Output> protocol,
+    Input input,
+  ) async {
+    var uri = baseUri;
+    var path = request.path;
+    final pattern = UriPattern.parse(path);
+    if (input is! HasLabel) {
+      if (pattern.labels.isNotEmpty) {
+        throw MissingLabelException(input, pattern.labels.first.toString());
+      }
+    } else {
+      path = expandLabels(path, input.labelFor);
+    }
+    var needsTrailingSlash = request.path.split('?').first.endsWith('/');
+    if (path.startsWith('/')) {
+      path = path.substring(1);
+    }
+    final headers = {
+      ...protocol.headers,
+      ...request.headers.asMap(),
+    };
+    final Map<String, Iterable<String>> queryParameters = {
+      for (final literal in pattern.queryLiterals.entries)
+        literal.key: [literal.value],
+      ...request.queryParameters.asMap(),
+    };
+    final body = protocol.serialize(input, specifiedType: FullType(Input));
+    var host = uri.host;
+    if (!endpoint.isHostnameImmutable && request.hostPrefix != null) {
+      var prefix = request.hostPrefix!;
+      if (input is HasLabel) {
+        prefix = expandHostLabel(prefix, input.labelFor);
+      }
+      host = '$prefix$host';
+    }
```

Comment:
There are a lot of variables in here that could be final with some refactoring. It makes it a little easier to read.

```suggestion
    final prefix = _createPrefix(
      endpoint: endpoint,
      request: request,
      input: input,
    );
    final host = '$prefix${uri.host}';
```

```dart
  String? _createPrefix({
    required Endpoint endpoint,
    required HttpRequest request,
    required Input input,
  }) {
    if (!endpoint.isHostnameImmutable && request.hostPrefix != null) {
      if (input is HasLabel) {
        return expandHostLabel(request.hostPrefix!, input.labelFor);
      }
      return request.hostPrefix;
    }
    return null;
  }
```